kolibri/core/assets/src/views/k-textbox/index.vue

```diff
@@ -125,10 +120,13 @@
       },
       emitKeydown(e) {
         /**
-         * Emits keydown event with event
+         * Emits keydown event
          */
         this.$emit('keydown', e);
       },
+      emitBlur() {
+        this.blurTimeout = window.setTimeout(() => this.$emit('blur'), 100);
```

Comment:
why wait 100ms to fire?

Reply:
Currently, as soon as a field in invalid, the submit button is disabled. So if the first field is focused, and then the submit button is clicked, the blur event is fired first and as a result, the submit button becomes disabled so a click event is not emitted and the form is not submitted. By delaying the blur, we can trigger the click event and submit the form before the button becomes disabled.

Reply:
Hm okay. As a general rule, introducing timing offsets to try and affect order of operations is brittle, and liable to introduce subtle bugs down the line.

It sounds like this stems from the [form spec](https://docs.google.com/document/d/1rrq8fovV5wBtUKv65wfRqXxP4SfnV30anUZHfzCVmN4/edit) which says that the buttons should be enabled unless there is a visible error shown? So, even though the field contents are invalid, the user hasn't seen an error yet and therefore we allow submission? It's a subtle situation, which makes me slightly want to change the spec'd behavior. But for now let's leave it since it's well-defined.

There are two things I can think of doing to make this a little cleaner while leaving the logic basically the same:

* Try switching from a timeout to a vue `nextTick`. While still a code smell, it's a little less arbitrary than 100ms
* Since this logic is specific to form handling patterns, move the delay out of the textbox component and up to the form level, in the event listener. This makes it less likely that a future user of textboxes will encounter subtle bugs because of an unexpected internal event firing delay

Reply:
> So, even though the field contents are invalid, the user hasn't seen an error yet and therefore we allow submission? 

Exactly.

> Try switching from a timeout to a vue nextTick. While still a code smell, it's a little less arbitrary than 100ms

Tried but blur is still fired before click.

Reply:
> Since this logic is specific to form handling patterns, move the delay out of the textbox component and up to the form level, in the event listener. 

Agreed.


@indirectlylit Actually, we might want to change the spec since there is an inconsistency since the click event of a submit button fires on mouseup, which can be well after 100ms. So two options are:
1. Start validation on focus.
2. Do not disable the submit button.

Reply:
Just to chip in here and concur that this sort of timeout will produce different kinds of unexpected behaviour (not least because other processing could cause the actual execution of the timeout to be significantly later). If the spec simply isn't implementable without this, I would strongly recommend changing the spec.

Reply:
There has been a fair amount of back-and-forth on whether or not it's even desirable to disable the submit button, so I'm inclined to lean towards (2) with a slight amendment to the 'Submission' section of the forms spec:

* Remove: _If there is a field that is clearly marked with an error on the client side, the submit button for the form should be disabled until the client side errors on each affected field are gone._
* Add: _If the submit button is clicked and there are visible client-side errors, set keyboard focus in the first field that contains an error_

If this sounds good to you, we can run it by the designers.

Reply:
I agree. Generally disabling submit buttons is not recommended. Sounds like a better pattern. :+1: 

To focus on the textbox, I would have to do something like `this.$refs.password.$el.querySelector('input').focus();`

Reply:
> To focus on the textbox, I would have to do something like `this.$refs.password.$el.querySelector('input').focus();`

This is one of the few cases where I think it's reasonable to give the textbox an imperative, parent-facing `focus` method.

Reply:
:sweat_smile: Sounds good.

Reply:
@indirectlylit We are going to have to document this in the style guide.

Reply:
@indirectlylit Added support for method documentation. Implemented the requested changes.